### PR TITLE
Fix #1951: Optimise hints_summary.xml

### LINKED ISSUE #1951: Optimise hints_summary.xml
This issue is tracking the optimisation work of `hints_summary.xml`

1. Make sure that the layout uses `ConstraintLayout` and there is minimal nesting of views. If this view already uses `ConstraintLayout` then it's fine.

2. Make sure that all textviews use the style from `styles.xml` file like `Heading1`, `TextFieldLabel`, `Body`, etc. whichever suits best.

While making the above changes the UI should not get affected and the end result should still be same as before (or an optimised version of UI).

If there are multiple files with the same name, update them all.

## References
Mobile Portrait: https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/grid/
Mobile Landscape: https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/grid/
Tablet: https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/grid/
Wiki: https://github.com/oppia/oppia-android/wiki/Working-on-UI

### PR DESCRIPTION
**Pull request by secjedi**
- Fix #1951
- Changed the outer Linear Layout to Constraint Layout
- Adapted the styles from the styles.xml file into hint_summary.xml
- Reviewer: @BenHenning 

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Review by rt4914 — CHANGES_REQUESTED** [review]
In this PR, you have done good with change related to constraint layout but all the margins have been disturbed. These changes needs to be done in such a way that before and after UI looks exactly the same.

But you can see below screenshot that the UI has changed.

Screenshot_1602593461
Screenshot_1602593611

Also, I am assuming that you did not check this UI by running the app and therefore I recommend you to ask questions in such cases.
**How to test UI of hints**
1. Go to Home Screen
2. Click on `Fractions` topic.
3. Click on `Lessons`
4. Click on `Matthew Goes to Bakery`
5. Select `What is a Fraction`
6. Click on continue until you reach first question.
7. In first question answer `One/Fifth` twice.
8. At bottom a blue box will appear, click on it and you will see hints.

**Comment by rt4914** [comment]
@secjedi PTAL

**Comment by secjedi** [comment]
Ok, I'm on it now. Thanks for the feedback

> @secjedi PTAL

**Comment by rt4914** [comment]
I don't know if its is ready for review or not but at this point it is not giving correct results.

Screenshot 2020-10-15 at 3 07 48 AM
Screenshot_1602711237



On think that you should focus on is changing the UI also making sure that margin/padding remain same.

Maybe these can help:
https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/grid/

https://github.com/oppia/oppia-android/wiki/Working-on-UI

**Comment by secjedi** [comment]
> I don't know if its is ready for review or not but at this point it is not giving correct results.
> 
> Screenshot 2020-10-15 at 3 07 48 AM
> 
> Screenshot_1602711237
> On think that you should focus on is changing the UI also making sure that margin/padding remain same.
> 
> Maybe these can help:
> https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/grid/
> 
> https://github.com/oppia/oppia-android/wiki/Working-on-UI

Thanks @rt4914 I am fixing it now

**Comment by secjedi** [comment]
> > I don't know if its is ready for review or not but at this point it is not giving correct results.
> > Screenshot 2020-10-15 at 3 07 48 AM
> > Screenshot_1602711237
> > On think that you should focus on is changing the UI also making sure that margin/padding remain same.
> > Maybe these can help:
> > https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/grid/
> > https://github.com/oppia/oppia-android/wiki/Working-on-UI
> 
> Thanks @rt4914 I am fixing it now

@rt4914 Please could you review the recent changes? Thanks

**Comment by rt4914** [comment]
@secjedi As our conversation please update the PR such that the UI remains correct and only the code is optimised.

**Comment by secjedi** [comment]
> @secjedi As our conversation please update the PR such that the UI remains correct and only the code is optimised.

@rt4914 Thank you for the walkthrough. Now, upon using the exact dimension from the develop branch and also from the design, the bottom padding is lost. You can view it here: https://res.cloudinary.com/dsxf1ke0u/image/upload/v1602930586/Screenshot_20201017-122756_yaxtum.jpg
So I do not want to manipulate further without informing you.

**Comment by rt4914** [comment]
@secjedi you can add bottom margin but I cannot confirm about the exact solution until I see the code, can you please push the latest changes that you have done.

**Comment by secjedi** [comment]
> @secjedi you can add bottom margin but I cannot confirm about the exact solution until I see the code, can you please push the latest changes that you have done.

Alright @rt4914

**Comment by secjedi** [comment]
> > @secjedi you can add bottom margin but I cannot confirm about the exact solution until I see the code, can you please push the latest changes that you have done.
> 
> Alright @rt4914

@rt4914 PTAL Please could you check my recent changes? I used the dimensions from the original file and the design.

**Comment by rt4914** [comment]
@secjedi Actually at this point its really hard to check the correctness of your PR as the `Files Changed` section shows changes in almost 49 files. I know you have not done changes to those files.

You can fix this by updating your `hints_summary` branch with latest `oppia-android/develop` branch.

PTAL:
Screenshot 2020-10-22 at 8 32 50 PM

**Comment by secjedi** [comment]
> @secjedi Actually at this point its really hard to check the correctness of your PR as the `Files Changed` section shows changes in almost 49 files. I know you have not done changes to those files.
> 
> You can fix this by updating your `hints_summary` branch with latest `oppia-android/develop` branch.
> 
> PTAL:
> Screenshot 2020-10-22 at 8 32 50 PM

@rt4914 I do have issues pushing old changes along side new ones> I've been trying to fix this for a while not. But I think you could review  
4666f3e

**Comment by secjedi** [comment]
> > @secjedi Actually at this point its really hard to check the correctness of your PR as the `Files Changed` section shows changes in almost 49 files. I know you have not done changes to those files.
> > You can fix this by updating your `hints_summary` branch with latest `oppia-android/develop` branch.
> > PTAL:
> > Screenshot 2020-10-22 at 8 32 50 PM
> 
> @rt4914 I do have issues pushing old changes along side new ones> I've been trying to fix this for a while not. But I think you could review
> [4666f3e](https://github.com/oppia/oppia-android/commit/4666f3e625d0b2f4fb84e5cfe21efe42b979ab56)
> @rt4914 @MohamedMedhat1998 @BenHenning  Please could you review my last commit? d684b52
Thank you

File: .idea/codeStyles/Project.xml [hunk 1]
```diff
@@ -161,4 +180,4 @@
       </indentOptions>
     </codeStyleSettings>
   </code_scheme>
-</component>
+</component>
```
Comment (rt4914) [inline .idea/codeStyles/Project.xml]: @secjedi These changes needs to be reverted. This was done unintentionally by you and therefore we should revert it.

File: app/src/main/res/layout/hints_summary.xml [hunk 2]
```diff
@@ -95,21 +94,20 @@
...
-        android:fontFamily="sans-serif"
-        android:textColor="@color/oppiaPrimaryText"
-        android:textSize="16sp" />
+        style="@style/Body"
+        android:textColor="@color/oppiaPrimaryText" />
     </LinearLayout>
-  </LinearLayout>
+  </androidx.constraintlayout.widget.ConstraintLayout>
 </layout>
+
```
Comment (rt4914) [inline app/src/main/res/layout/hints_summary.xml:113]: Remove this extra blank line.

File: app/src/main/res/layout-sw600dp-port/hints_summary.xml [hunk 3]
```diff
@@ -90,19 +87,18 @@
...
-        android:fontFamily="sans-serif"
-        android:textColor="@color/oppiaPrimaryText"
-        android:textSize="16sp" />
+        style="@style/Body"
+        android:textColor="@color/oppiaPrimaryText" />
     </LinearLayout>
-  </LinearLayout>
-</layout>
+  </androidx.constraintlayout.widget.ConstraintLayout>
+</layout>
```
Comment (rt4914) [inline app/src/main/res/layout-sw600dp-port/hints_summary.xml]: EOF is missing, add one blank line at the end.

File: app/src/main/res/layout-sw600dp-land/hints_summary.xml [hunk 4]
```diff
@@ -91,18 +89,17 @@
...
         android:layout_gravity="start"
-        android:fontFamily="sans-serif"
         android:textColor="@color/oppiaPrimaryText"
-        android:textSize="16sp" />
+        style="@style/Body" />
     </LinearLayout>
-  </LinearLayout>
-</layout>
+  </androidx.constraintlayout.widget.ConstraintLayout>
+</layout>
```
Comment (rt4914) [inline app/src/main/res/layout-sw600dp-land/hints_summary.xml]: EOF is missing, add one blank line at the end.

File: app/src/main/res/layout-land/hints_summary.xml [hunk 5]
```diff
@@ -10,106 +11,101 @@
...
-      </LinearLayout>
+      <TextView
+        android:id="@+id/hints_and_solution_summary"
+        style="@style/Body"
+        android:layout_gravity="start"
+        android:textColor="@color/oppiaPrimaryText" />
     </LinearLayout>
-</layout>
+  </androidx.constraintlayout.widget.ConstraintLayout>
+</layout>
```
Comment (rt4914) [inline app/src/main/res/layout-land/hints_summary.xml]: EOF is missing, add one blank line at the end.

**Review by rt4914 — CHANGES_REQUESTED** [review]
Suggested minor changes. I will still have to take detail look tomorrow.

**Comment by rt4914** [comment]
@secjedi PTAL

**Comment by secjedi** [comment]
> @secjedi PTAL

@rt4914 Please kindly review 41e5c07

**Comment by rt4914** [comment]
> > @secjedi PTAL
> 
> @rt4914 Please kindly review [41e5c07](https://github.com/oppia/oppia-android/commit/41e5c077356804c2a25e2f96efd12858196310e7)

@secjedi Your last commit solves couple of issues but not all the issues which I have mentioned above.

Also, please reply to all the comments before I take another pass on this.

File: .idea/codeStyles/Project.xml (hunk 1 above)
Comment (secjedi) [reply .idea/codeStyles/Project.xml]: I reverted this already

File: app/src/main/res/layout/hints_summary.xml (hunk 2 above)
Comment (secjedi) [reply app/src/main/res/layout/hints_summary.xml:113]: I removed the extra line. Now the file ends at line 112

File: app/src/main/res/layout-sw600dp-port/hints_summary.xml (hunk 3 above)
Comment (secjedi) [reply app/src/main/res/layout-sw600dp-port/hints_summary.xml]: I added an extra line here. File now ends at line 105

File: app/src/main/res/layout-sw600dp-land/hints_summary.xml (hunk 4 above)
Comment (secjedi) [reply app/src/main/res/layout-sw600dp-land/hints_summary.xml]: Added an extra line here. File now ends at line 106

File: app/src/main/res/layout-land/hints_summary.xml (hunk 5 above)
Comment (secjedi) [reply app/src/main/res/layout-land/hints_summary.xml]: Also added an extra line here. File now ends at line 112

**Comment by secjedi** [comment]
> > > @secjedi PTAL
> > 
> > 
> > @rt4914 Please kindly review [41e5c07](https://github.com/oppia/oppia-android/commit/41e5c077356804c2a25e2f96efd12858196310e7)
> 
> @secjedi Your last commit solves couple of issues but not all the issues which I have mentioned above.
> 
> Also, please reply to all the comments before I take another pass on this.

@rt4914 I made the necessary additions and subtractions as suggested. Please do look closely at the line numbers. See 3049407

File: app/src/main/res/layout-land/hints_summary.xml [hunk 6]
```diff
@@ -1,4 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
+
```
Comment (rt4914) [inline app/src/main/res/layout-land/hints_summary.xml:2]: Remove this extra line.

File: app/src/main/res/layout-land/hints_summary.xml [hunk 7]
```diff
@@ -10,106 +11,102 @@
...
-      </LinearLayout>
+      <TextView
+        android:id="@+id/hints_and_solution_summary"
+        style="@style/Body"
+        android:layout_gravity="start"
+        android:textColor="@color/oppiaPrimaryText" />
     </LinearLayout>
+  </androidx.constraintlayout.widget.ConstraintLayout>
 </layout>
+
```
Comment (rt4914) [inline app/src/main/res/layout-land/hints_summary.xml:112]: Remove this extra line.

File: app/src/main/res/layout-sw600dp-land/hints_summary.xml [hunk 8]
```diff
@@ -91,18 +89,18 @@
...
         android:layout_gravity="start"
-        android:fontFamily="sans-serif"
         android:textColor="@color/oppiaPrimaryText"
-        android:textSize="16sp" />
+        style="@style/Body" />
     </LinearLayout>
-  </LinearLayout>
+  </androidx.constraintlayout.widget.ConstraintLayout>
 </layout>
+
```
Comment (rt4914) [inline app/src/main/res/layout-sw600dp-land/hints_summary.xml:106]: Remove this extra line.

File: app/src/main/res/layout-sw600dp-port/hints_summary.xml [hunk 9]
```diff
@@ -90,19 +87,19 @@
...
-        android:fontFamily="sans-serif"
-        android:textColor="@color/oppiaPrimaryText"
-        android:textSize="16sp" />
+        style="@style/Body"
+        android:textColor="@color/oppiaPrimaryText" />
     </LinearLayout>
-  </LinearLayout>
+  </androidx.constraintlayout.widget.ConstraintLayout>
 </layout>
+
```
Comment (rt4914) [inline app/src/main/res/layout-sw600dp-port/hints_summary.xml:105]: Remove this extra line.

File: app/src/main/res/layout/hints_summary.xml [hunk 10]
```diff
@@ -1,4 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
+
```
Comment (rt4914) [inline app/src/main/res/layout/hints_summary.xml:2]: Remove this extra line.

File: app/src/main/res/layout/hints_summary.xml (hunk 2 above)
Comment (rt4914) [reply app/src/main/res/layout/hints_summary.xml:113]: Actually that extra blank line is still there.

File: .idea/codeStyles/Project.xml [hunk 11]
```diff
@@ -5,10 +5,29 @@
...
+          <package name="" alias="true" withSubpackages="true" />
+        </value>
+      </option>
       <option name="NAME_COUNT_TO_USE_STAR_IMPORT" value="2147483647" />
       <option name="NAME_COUNT_TO_USE_STAR_IMPORT_FOR_MEMBERS" value="2147483647" />
       <option name="CODE_STYLE_DEFAULTS" value="KOTLIN_OFFICIAL" />
     </JetCodeStyleSettings>
+    <MarkdownNavigatorCodeStyleSettings>
+      <option name="RIGHT_MARGIN" value="72" />
+    </MarkdownNavigatorCodeStyleSettings>
```
Comment (rt4914) [inline .idea/codeStyles/Project.xml:30]: Please remove these changes.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@secjedi 

Your PR still has following issues:

1. Extra blank lines either at the end of file or at some other places. We should not have blank lines at random places and at the end of file just after the last character there should be only one blank line.
Something like this:
Screenshot 2020-10-29 at 4 42 45 PM


2. Also there are changes to `Project.xml` file. We cannot merge a PR with changes in this PR and therefore you will need to revert that file changes.

Screenshot 2020-10-29 at 4 39 51 PM


Thanks @secjedi

**Comment by anandwana001** [comment]
@secjedi any update on this?

**Comment by anandwana001** [comment]
Closing this PR as no further update.

**Pull request closed without merging**